{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5d96024-64fc-435e-822e-c028e52f255b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import inspect\n",
    "from jw_utils import jw_draw_tree as jdt\n",
    "from ete3 import ncbi_taxonomy\n",
    "from ete3 import Tree\n",
    "ncbi_tax = ncbi_taxonomy.NCBITaxa()\n",
    "import json\n",
    "from jw_utils import ncbi_datasets_fxs as nfx\n",
    "from orthofinder_utils import dash_ortho_parser as dop\n",
    "from Bio import Phylo\n",
    "from plotly import graph_objects as go\n",
    "from jw_utils import jw_ncbi_taxonomy as jnt\n",
    "import os\n",
    "from jw_utils import plotly_preferences as pprefs\n",
    "pprefs.__file__\n",
    "import pandas as pd\n",
    "from orthofinder_utils import itol_annotations as itol_annots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7ae9118-1555-4baa-80a7-16ed90dd93df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "rgb_to_hex = {'rgba(132,137,145,1)':'#848991',\n",
    " 'rgb(0,208,132)':'#00d084',\n",
    " 'rgb(0,122,255)':'#007aff',\n",
    " 'rgb(171,184,195)':'#abb8c3',\n",
    " 'rgb(255,105,0)':'#ff6900',\n",
    " 'rgb(252,185,0)':'#fcb900',\n",
    " 'rgb(123,220,181)':'#7bdcb5',\n",
    " 'rgb(142,209,252)':'#8ed1fc',\n",
    " 'rgb(6,147,227)':'#0693e3',\n",
    " 'rgb(155,81,224)':'#9b51e0',\n",
    " 'rgb(207,46,46)':'#cf2e2e',\n",
    " 'rgb(247,141,167)':'#f78da7'}\n",
    "\n",
    "hex_colors = list(rgb_to_hex.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0eeee150-a3cc-40e2-b8ef-730359de48e9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('./all_strains_summary.tsv', sep='\\t')[['Strain ID', 'Full organism name', 'Species',\n",
    "       'Strain', 'Assembly level', 'Host taxonomy ID', 'Host Disease', \n",
    "       'Isolation Source', 'Assembly name', 'Refseq assembly accession',\n",
    "       'Scaffold N50', 'Contig N50']]\n",
    "df_disease = df.loc[df['Host Disease']=='cystic fibrosis',:]\n",
    "df_disease = df_disease.loc[df['Assembly level']=='Complete Genome',:]\n",
    "CF_accessions = list(df_disease['Refseq assembly accession'])\n",
    "df_disease = df_disease.set_index('Refseq assembly accession')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "34ec75ef-e2c7-4630-9225-3ce0c3e5bae6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOGs = {\n",
    "    'crc':'N3.HOG0004558', 'cbrA':'N3.HOG0005021', 'pelA':'N3.HOG0009322','pelB':'N3.HOG0009282','pelC':'N3.HOG0009225','cupB1':'N3.HOG0006832',\n",
    "    'cupB2':'N3.HOG0001264','cupB3':'N3.HOG0001773','cupC1':'N3.HOG0007291','cupC2':'N3.HOG0001265','cupC3':'N3.HOG0001776',\n",
    "    }\n",
    "HOGs_truncated = {\n",
    "    'cbrA':'N3.HOG0005021', 'crc':'N3.HOG0004558', 'pelA':'N3.HOG0009322','cupB1':'N3.HOG0006832','cupC2':'N3.HOG0001265',\n",
    "     }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30bcd77b-71c5-4637-af54-846876a2afc5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "CF_accessions\n",
    "with open('./CF_accessions.txt', 'w') as f:\n",
    "    for acc in CF_accessions:\n",
    "        f.write(acc + '\\n')\n",
    "#!datasets download genome accession --inputfile ./CF_accessions.txt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d3ec725b-f08b-4670-869d-c25d7a2703bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dop_obj = dop.DashOrthoParser('./data', tax_level='N3')\n",
    "tree =  Phylo.read('./data/Species_Tree/SpeciesTree_rooted_node_labels.txt', format='newick')\n",
    "HOGs_to_name_d = {k:v for v,k in HOGs_truncated.items()}\n",
    "names, binary_count_d = itol_annots.make_binary_count_dict(tree, HOGs_truncated.values(),dop_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "26626de5-5235-41dc-8041-0c5a273f9d65",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_002362295.1,0,0,0,0,0\n",
      "GCF_000046845.1,0,0,0,0,0\n",
      "GCF_005281455.1,0,0,0,0,0\n",
      "GCA_001077675.1,0,0,0,0,0\n",
      "GCF_000413935.1,0,0,0,0,0\n",
      "GCF_001682515.1,0,0,0,0,0\n",
      "GCF_900105655.1,1,1,0,0,0\n",
      "GCF_900105255.1,1,1,1,0,0\n",
      "GCF_900105005.1,1,1,0,0,0\n",
      "GCF_009497895.1,1,1,0,0,0\n",
      "GCF_020025155.1,1,1,1,0,0\n",
      "GCF_900104945.1,1,1,1,0,0\n",
      "GCF_008693965.1,1,1,0,0,0\n",
      "GCF_014522265.1,1,1,1,0,0\n",
      "GCF_001518815.1,1,1,1,0,0\n",
      "GCF_900105995.1,1,1,0,0,0\n",
      "GCF_900105885.1,1,1,0,0,0\n",
      "GCF_900104805.1,1,1,0,0,0\n",
      "GCF_024448735.1,1,1,1,0,0\n",
      "GCF_021432765.1,1,1,0,0,0\n",
      "GCF_009740185.1,1,1,0,0,0\n",
      "GCF_000327065.1,1,1,0,0,0\n",
      "GCF_000825645.2,1,1,0,0,0\n",
      "GCF_900636845.1,1,1,0,0,0\n",
      "GCF_000013785.1,1,1,0,0,0\n",
      "GCF_000219605.1,1,1,0,0,0\n",
      "GCF_021605025.1,1,1,0,1,0\n",
      "GCF_019879085.1,1,1,0,1,0\n",
      "GCF_011397855.1,1,1,0,1,0\n",
      "GCF_020422905.1,1,1,0,1,0\n",
      "GCF_013393345.1,1,1,0,1,1\n",
      "GCF_001708505.1,1,1,0,0,1\n",
      "GCF_009911755.1,1,1,0,1,0\n",
      "GCF_004745455.1,1,1,0,1,0\n",
      "GCF_000349845.1,1,1,0,1,0\n",
      "GCF_900188285.1,1,1,0,0,0\n",
      "GCF_012971725.1,1,1,0,0,0\n",
      "GCF_000017205.1,1,1,1,1,1\n",
      "GCF_022559545.1,1,1,1,1,1\n",
      "GCF_013343535.1,1,1,1,1,1\n",
      "GCF_013343475.1,1,1,0,1,1\n",
      "GCF_013343435.1,1,1,1,1,1\n",
      "GCF_004355125.1,1,1,1,1,1\n",
      "GCF_004355145.1,1,1,1,1,1\n",
      "GCF_000510305.1,1,1,1,1,1\n",
      "GCF_000271365.1,1,1,1,1,1\n",
      "GCF_000568855.2,1,1,1,1,1\n",
      "GCF_011466835.1,1,1,1,1,1\n",
      "GCF_000026645.1,1,1,1,1,1\n",
      "GCF_000006765.1,1,1,1,1,1\n",
      "GCA_000014625.1,1,1,1,1,1\n",
      "GCF_013200955.1,1,1,0,0,0\n",
      "GCF_016861485.1,1,1,0,0,0\n",
      "GCF_021654315.1,1,1,0,0,0\n",
      "GCF_001839655.2,1,1,0,0,0\n",
      "GCF_024397895.1,1,1,1,0,0\n",
      "GCF_900116605.1,1,1,1,1,0\n",
      "GCF_900682675.1,1,1,1,1,0\n",
      "GCF_017915135.1,1,1,1,0,0\n",
      "GCF_900626055.1,1,1,1,0,0\n",
      "GCF_002442555.1,1,1,0,0,0\n",
      "GCF_021495345.1,1,1,0,0,0\n",
      "GCF_017798225.1,1,1,0,0,0\n",
      "GCF_000204295.1,1,1,0,0,0\n",
      "GCF_021029445.1,1,1,0,0,0\n",
      "GCF_002941105.1,1,1,0,0,0\n",
      "GCF_013409125.2,1,1,0,0,0\n",
      "GCF_002741095.1,1,1,1,0,0\n",
      "GCF_900106015.1,1,1,0,0,0\n",
      "GCF_015712065.1,1,1,1,0,0\n",
      "GCF_000319815.1,1,1,1,0,0\n",
      "GCF_021601925.1,1,1,0,0,0\n",
      "GCF_000696345.2,1,1,0,1,1\n",
      "GCF_009755645.1,1,1,0,1,0\n",
      "GCF_000746525.1,1,1,0,1,1\n",
      "GCF_000759535.1,1,1,0,0,0\n",
      "GCF_019139895.1,1,1,0,0,1\n",
      "GCF_019139835.1,1,1,0,0,0\n",
      "GCF_003940825.1,1,1,0,1,0\n",
      "GCF_023277925.1,1,1,0,0,1\n",
      "GCF_000026105.1,1,1,0,0,0\n",
      "GCF_019139535.1,1,1,0,0,1\n",
      "GCF_014268885.2,1,1,0,0,1\n",
      "GCF_002736065.1,1,1,0,0,1\n",
      "GCF_001515585.1,1,1,0,0,0\n",
      "GCF_002884415.1,1,1,0,0,0\n",
      "GCF_019139675.1,1,1,0,0,0\n",
      "GCF_005080685.1,1,1,0,0,0\n",
      "GCF_014269025.1,1,1,0,0,0\n",
      "GCF_024023035.1,1,1,0,0,0\n",
      "GCF_021725455.1,1,1,0,0,0\n",
      "GCF_001477425.1,1,1,0,0,0\n",
      "GCF_002951475.1,1,1,0,0,0\n",
      "GCF_002688705.1,0,1,0,0,0\n",
      "GCF_017166425.1,1,1,0,0,0\n",
      "GCF_007833495.1,1,1,0,0,0\n",
      "GCF_001655295.1,1,1,0,0,1\n",
      "GCF_013425825.1,1,1,0,0,1\n",
      "GCF_002943995.1,1,1,0,0,0\n",
      "GCF_014269225.2,1,1,0,1,1\n",
      "GCF_022049085.1,1,1,0,0,0\n",
      "GCF_014656565.1,1,1,0,0,0\n",
      "GCF_002864795.1,1,1,0,0,0\n",
      "GCF_000510285.1,1,1,0,0,0\n",
      "GCF_000761155.1,1,1,0,0,0\n",
      "GCF_013374995.1,1,1,0,1,0\n",
      "GCF_003415775.1,1,1,0,0,1\n",
      "GCF_000452745.2,1,1,0,0,1\n",
      "GCF_018388545.1,1,1,0,0,1\n",
      "GCF_024413275.1,1,1,0,0,1\n",
      "GCF_003699955.1,1,1,0,0,1\n",
      "GCF_011765175.1,1,1,0,0,1\n",
      "GCF_000452705.1,1,1,0,0,1\n",
      "GCF_016599635.1,1,1,0,0,1\n",
      "GCF_000145845.1,1,1,0,0,1\n",
      "GCF_000145845.2,1,1,0,0,1\n",
      "GCF_002905795.1,1,1,0,0,1\n",
      "GCF_009903285.1,1,0,0,0,1\n",
      "GCF_000344475.3,1,1,0,0,1\n",
      "GCF_002024285.1,1,1,0,0,1\n",
      "GCF_003700835.1,1,1,0,0,1\n",
      "GCF_002966555.1,1,1,0,0,1\n",
      "GCF_000172895.1,1,1,0,0,1\n",
      "GCF_000007805.1,1,1,0,0,1\n",
      "GCF_000145885.1,1,1,0,0,1\n",
      "GCF_900235895.1,1,1,0,0,1\n",
      "GCF_002318885.1,1,1,0,0,1\n",
      "GCF_000988485.1,1,1,0,0,1\n",
      "GCF_022510205.1,1,1,0,0,1\n",
      "GCF_003047185.1,1,1,0,0,1\n",
      "GCF_000452445.2,1,1,0,0,1\n",
      "GCF_001400495.1,1,1,0,0,1\n",
      "GCF_001482725.1,1,1,0,0,1\n",
      "GCF_022557235.1,1,1,0,0,1\n",
      "GCF_020917325.1,1,1,0,0,1\n",
      "GCF_019334485.1,1,1,0,0,1\n",
      "GCF_023207855.1,1,1,0,0,1\n",
      "GCF_022026035.1,1,1,0,0,1\n",
      "GCF_020204845.1,1,1,0,0,1\n",
      "GCF_014855475.1,1,1,0,0,1\n",
      "GCF_013911125.1,1,1,0,0,1\n",
      "GCF_003413075.1,0,1,0,0,1\n",
      "GCF_003699405.1,1,1,0,0,1\n",
      "GCF_001400675.1,1,1,0,0,1\n",
      "GCF_000164015.3,1,1,0,0,1\n",
      "GCF_000146005.2,1,1,0,0,1\n",
      "GCF_001020725.2,1,1,1,0,0\n",
      "GCF_900475285.1,1,1,0,1,0\n",
      "GCF_011040435.1,1,1,0,0,1\n",
      "GCF_015476275.1,1,1,0,0,0\n",
      "GCF_001543125.1,1,1,1,1,1\n",
      "GCF_000280575.1,1,1,0,1,1\n",
      "GCF_014268755.2,1,1,0,1,1\n",
      "GCF_013403585.1,1,1,0,0,1\n",
      "GCF_900106045.1,1,1,0,1,0\n",
      "GCF_014268455.2,1,1,0,1,1\n",
      "GCF_001641345.1,0,0,0,1,1\n",
      "GCF_003626995.1,1,1,0,1,1\n",
      "GCF_004124255.1,0,0,0,0,0\n",
      "GCF_001647715.1,1,1,0,1,1\n",
      "GCF_900475215.1,1,1,0,0,1\n",
      "GCF_900625045.1,1,1,1,1,0\n",
      "GCF_021233455.1,1,1,0,1,1\n",
      "GCF_002980135.1,1,1,1,1,0\n",
      "GCF_018417575.1,1,1,0,1,0\n",
      "GCF_001730655.1,1,1,0,1,0\n",
      "GCF_900104915.1,1,1,0,1,0\n",
      "GCF_022807995.1,1,1,0,0,1\n",
      "GCF_019139815.1,1,1,0,1,0\n",
      "GCF_013778535.1,1,1,0,1,0\n",
      "GCF_900601635.1,1,1,0,1,0\n",
      "GCF_019285655.1,1,1,0,0,0\n",
      "GCF_022870785.1,1,1,0,1,1\n",
      "GCF_014268375.2,1,1,0,1,0\n",
      "GCF_901482615.1,1,1,0,1,0\n",
      "GCF_003852025.1,1,1,0,1,1\n",
      "GCF_001297075.1,1,1,0,1,1\n",
      "GCF_014268275.2,1,1,1,0,0\n",
      "GCF_016134935.1,0,1,1,0,1\n",
      "GCF_002966835.1,1,1,1,1,1\n",
      "GCF_000009225.2,1,1,0,0,0\n",
      "GCF_008801565.1,1,1,1,0,1\n",
      "GCF_022533565.1,1,1,1,1,0\n",
      "GCF_002007785.1,1,1,1,1,0\n",
      "GCF_019139855.1,1,1,0,1,1\n",
      "GCF_001547895.1,1,1,1,1,0\n",
      "GCF_001547915.1,1,1,1,1,0\n",
      "GCF_000828695.1,1,1,1,1,1\n",
      "GCF_000397205.1,1,1,1,1,1\n",
      "GCF_000012265.1,1,1,1,1,1\n",
      "GCF_003851145.1,1,1,0,1,1\n",
      "GCF_003850425.1,1,1,0,1,1\n",
      "GCF_003850385.1,1,1,0,1,1\n",
      "GCF_020424105.1,1,0,0,1,1\n",
      "GCF_900636995.1,1,1,0,1,1\n",
      "GCF_003851985.1,1,1,0,1,1\n",
      "GCF_009362955.1,1,1,0,1,1\n",
      "GCF_003851385.1,1,1,0,1,1\n",
      "GCF_000237045.1,1,1,0,1,1\n",
      "GCF_000813225.1,1,1,0,1,1\n",
      "GCF_014076455.1,1,1,0,0,1\n",
      "GCF_001412045.1,1,1,0,0,0\n",
      "GCF_001708425.1,1,1,0,0,0\n",
      "GCF_019141525.1,1,1,0,0,0\n",
      "GCF_013347965.1,1,1,0,0,0\n",
      "GCF_003410335.1,1,1,0,0,0\n",
      "GCF_001269655.1,1,1,0,0,0\n",
      "GCF_014268695.2,1,1,0,0,0\n",
      "GCF_021389915.1,1,1,0,0,0\n",
      "GCF_008370715.1,1,1,0,0,0\n",
      "GCF_000194805.1,1,1,0,0,0\n",
      "GCF_001449085.1,1,1,0,0,0\n",
      "GCF_004348895.1,1,1,0,1,0\n",
      "GCF_019139795.1,1,1,0,0,0\n",
      "GCF_002356535.2,1,1,0,0,0\n",
      "GCF_008801455.1,1,1,1,0,0\n",
      "GCF_001661075.1,1,1,0,0,0\n",
      "GCF_001042905.1,1,1,0,0,0\n",
      "GCF_014190825.1,1,1,0,0,0\n",
      "GCF_002355315.1,1,1,0,0,0\n",
      "GCF_001655615.1,1,1,0,0,0\n",
      "GCF_003671865.1,1,1,0,0,0\n",
      "GCF_000293885.2,1,1,1,0,0\n",
      "GCF_004135995.1,1,1,1,0,0\n",
      "GCF_002091695.1,1,1,1,0,0\n",
      "GCF_019693255.1,1,1,1,0,0\n",
      "GCF_014863605.1,1,1,0,0,0\n",
      "GCF_021398855.1,1,1,0,0,1\n",
      "GCF_014863585.1,1,1,0,0,1\n",
      "GCF_019719375.1,1,1,0,0,1\n",
      "GCF_001655595.1,1,1,0,0,0\n",
      "GCF_001594225.1,1,1,0,0,0\n",
      "GCF_014268495.1,1,1,0,0,1\n",
      "GCF_002287865.1,1,1,0,0,0\n",
      "GCF_004683905.1,1,1,0,0,0\n",
      "GCF_014268445.1,1,1,0,0,1\n",
      "GCF_014268485.2,1,1,0,0,1\n",
      "GCF_019614655.1,1,1,0,0,1\n",
      "GCF_900105485.1,1,1,0,0,0\n",
      "GCF_014268595.1,1,1,0,0,0\n",
      "GCF_024072295.1,1,1,0,0,1\n",
      "GCF_017167965.1,1,1,0,0,0\n",
      "GCF_014268585.2,1,1,0,0,1\n",
      "GCF_001605965.1,1,1,0,0,1\n"
     ]
    }
   ],
   "source": [
    "itol_outfile_path = './itol_tree/multibar_dataset_trunc.txt'\n",
    "itol_annots.make_multi_itol_bargraph_dataset('./itol_tree/multibar_dataset_trunc.txt',binary_count_d, names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b14a00b0-cab1-4d8f-b3d4-6791c554e314",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2bdb0a45-0f02-4150-b56e-f5d4353f26b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pseudomonas aeruginosa DK2\n",
      "Pseudomonas aeruginosa SCV20265\n",
      "Pseudomonas fulva FDAARGOS_167\n",
      "Pseudomonas monteilii FDAARGOS_171\n",
      "Pseudomonas aeruginosa AES1M\n",
      "Pseudomonas aeruginosa AES1R\n",
      "Pseudomonas aeruginosa DVT419\n",
      "Pseudomonas aeruginosa DVT779\n",
      "Pseudomonas aeruginosa DVT412\n",
      "Pseudomonas aeruginosa PES_P749\n"
     ]
    }
   ],
   "source": [
    "accessions_trunc = []\n",
    "for acc in CF_accessions:\n",
    "    if dop_obj.accession_to_name.get(acc):\n",
    "        print(dop_obj.accession_to_name.get(acc))\n",
    "        accessions_trunc.append(acc)\n",
    "df_disease_trunc = df_disease.loc[accessions_trunc, :][['Full organism name', 'Species', 'Strain', 'Assembly level', 'Host Disease', 'Isolation Source']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9d71c436-2efd-4cac-95bb-1691b8e436dd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_disease_trunc.to_csv('./Clinical_isolates.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "79a404b8-b283-435b-92ac-4da883c98c83",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "relabel_dict = {}\n",
    "for acc, name in dop_obj.accession_to_name.items():\n",
    "    name = name.replace('[','').replace(']','')\n",
    "    name_list =  name.split(' ')\n",
    "    genus = name_list[0][:1]+'.'\n",
    "    species = name_list[1]\n",
    "    #strain = ' '.join(name_list[2:])[:10]\n",
    "    strain = name_list[-1][:10]\n",
    "    name = f'{genus} {species} {strain}' \n",
    "    relabel_dict[acc]=name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3ed5b93d-a56e-499a-8225-339a0c955a46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "outfile_path = './itol_tree/rename_dataset.txt'\n",
    "itol_annots.relabel_itol_treeleafs(tree, relabel_dict,outfile_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9fcb5df7-d9f3-462f-9361-27193502d7a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "    \n",
    "outfile_path = './itol_tree/rename_dataset.txt'\n",
    "#relabel_itol_treeleafs(tree, relabel_dict, outfile_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "8b484bba-11a3-4a30-9efd-7698af09a434",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import shutil\n",
    "dest = './Proteomes'\n",
    "parent_fp = './data/ncbi_dataset/data/'\n",
    "for folder in [f for f in os.listdir(parent_fp) if f.startswith('GC')]:\n",
    "    proteome_fp = os.path.join(parent_fp,folder, 'sequence_report.jsonl')\n",
    "    if os.path.exists(proteome_fp):\n",
    "        n_dest_fp = os.path.join(dest, folder)\n",
    "        shutil.move(proteome_fp, n_dest_fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8da6a9a-6723-4dd2-a86d-5f5a60555a75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3ffb155b-b5dc-4225-bc8f-e0e3c539f39f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def make_simple_itol_bargraph_dataset(template_path, file_path, data_dict):\n",
    "    \"\"\"\n",
    "    Add data to simple bargraph template for itol dataset adn write to file.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    template_path : str\n",
    "        path to template file from itol.\n",
    "    file_path : str\n",
    "        path for new file that is to be created\n",
    "    data_dict :dict\n",
    "        {terminal node name: count}\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    template_lines = []\n",
    "    with open(template_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line[-1] != '\\n':\n",
    "                line = line+'\\n'\n",
    "            template_lines.append(line)\n",
    "            \n",
    "    with open(file_path, 'w') as f:\n",
    "        for line in template_lines:\n",
    "            f.write(line)\n",
    "        for name, count in data_dict.items():\n",
    "            f.write(f'{name},{count}\\n')\n",
    "            \n",
    "crc_data = list(copynumber_bargraph_data_dict([HOGs['crc']], dop_obj, id_type='accession').values())[0]\n",
    "template_path = './itol_tree/dataset_multibar_template.txt'\n",
    "file_path = './itol_tree/crc_simplebar_dataset.txt'\n",
    "make_simple_itol_bargraph_dataset(template_path, file_path, crc_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
